src/functions-and-arrays.js

```diff
@@ -106,9 +256,25 @@ const matrix = [
   [1, 70, 54, 71, 83, 51, 54, 69, 16, 92, 33, 48, 61, 43, 52, 1, 89, 19, 67, 48]
 ];
 
-function greatestProduct() {}
-
+function greatestProduct(arr) {
+  if (arr.length === 0){
+    return null;
+  }
+  if (arr.length === 1){
+    return arr[0];
+  }
+  let sum=0,x=arr.length,y=0;
+  for(let i=0;i<=arr.length-1;i++){
+    let arr2=arr[i];
+    y+=arr2.length;
+    for(let j=0;j<=arr2.length-1;j++){
+      sum+=arr2[j];
+    }
+  }
+  if (sum===y)return 1;
```

Comment:
this is passing the test cases but not really what the exercise asks for haha 

when you loop over the row, one thing you want to do is to loop only until row.length - 3; this way you can use the index j, j+1, j+2 and j+3 to get the 4 elements in the row without running over at the end with the j+3. then you multiply these values together to get one of the possible products for this row, push this product into a list of products and then find the maximum value of that list in the end... 

```javascript 
const allProducts = []
for (let row of arr) {
  for (let j = 0; j < row.length - 3; j++){
    allProducts.push(row[j] * row[j+1] * row[j+2] * row[j+3])
  }
}
return Math.max(...allProducts)
```

Reply:
the code I wrote there only covers the row case, but I leave it as an exercise to you to translate it to cover the column case as well :)